File: resizer/src/main/java/com/trailmagic/resizer/ImageResizeServiceImpl.java
package com.trailmagic.resizer;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

@Service
public class ImageResizeServiceImpl implements ImageResizeService {
    private ImageResizer imageResizer;

    @Autowired
    public ImageResizeServiceImpl(ImageResizer imageResizer) {
        this.imageResizer = imageResizer;
    }

    @SuppressWarnings({"ResultOfMethodCallIgnored"})
    @Override
    public List<ImageFileInfo> scheduleResize(InputStream srcInputStream) throws ResizeFailedException {
        File srcFile = writeFile(srcInputStream);
        List<ImageFileInfo> resultInfos = scheduleResize(srcFile);

        srcFile.delete();
        return resultInfos;
    }

    public List<ImageFileInfo> scheduleResize(File srcFile) {
        ImageFileInfo srcFileInfo = imageResizer.identify(srcFile);

        List<ImageFileInfo> resultInfos = new ArrayList<ImageFileInfo>();

        for (Integer size : Arrays.asList(128, 256, 512, 1024, 2048)) {
            if (size <= srcFileInfo.getWidth()) {
                resultInfos.add(resizeAndIdentify(srcFile, srcFileInfo, size));
            }
        }
        return resultInfos;
    }

    @Override
    public ImageFileInfo identify(File srcFile) {
        return imageResizer.identify(srcFile);
    }


    public File writeFile(InputStream srcInputStream) throws ResizeFailedException {
        try {
            return imageResizer.writeToTempFile(srcInputStream);
        } catch (IOException e) {
            throw new ResizeFailedException("Could not write src to temp file", e);
        }
    }

    private ImageFileInfo resizeAndIdentify(File srcFile, ImageFileInfo srcFileInfo, Integer size) throws ResizeFailedException {
        File file = imageResizer.resizeImage(srcFile, srcFileInfo, size);
        ImageFileInfo info = imageResizer.identify(file);
        info.setFile(file);
        return info;
    }
}

File: resizer/src/main/scala/com/trailmagic/resizer/ImageResizeServiceImpl.scala
package com.trailmagic.resizer

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.io.File
import java.io.IOException
import java.io.InputStream
import scala.collection.JavaConverters._

@Service
class ImageResizeServiceImpl @Autowired()(imageResizer: ImageResizer) extends ImageResizeService {
  def scheduleResize(srcInputStream: InputStream): java.util.List[ImageFileInfo] = {
    withTempFile(writeFile(srcInputStream)) {
      scheduleResize(_)
    }
  }

  def withTempFile[T](file: File)(block: (File) => T) = {
    try {
      block(file)
    } finally {
      file.delete
    }
  }

  def scheduleResize(srcFile: File): java.util.List[ImageFileInfo] = {
    val srcFileInfo: ImageFileInfo = imageResizer.identify(srcFile)
    val resultInfoList = List[ImageFileInfo]()
    val sizes = List(128, 256, 512, 1024, 2048)

    val newManifestations = sizes filter (_ <= srcFileInfo.getWidth) map (resizeAndIdentify(srcFile, srcFileInfo, _))
    newManifestations.asJava
  }

  def identify(srcFile: File): ImageFileInfo = {
    imageResizer.identify(srcFile)
  }

  def writeFile(srcInputStream: InputStream): File = {
    try {
      imageResizer.writeToTempFile(srcInputStream)
    }
    catch {
      case e: IOException => {
        throw new ResizeFailedException("Could not write src to temp file", e)
      }
    }
  }

  private def resizeAndIdentify(srcFile: File, srcFileInfo: ImageFileInfo, size: Int): ImageFileInfo = {
    val file = imageResizer.resizeImage(srcFile, srcFileInfo, size)
    val info = imageResizer.identify(file)
    info.setFile(file)
    info
  }
}

